tests/queries/test_qs_combinators.py

```diff
@@ -123,6 +123,33 @@ def test_union_nested(self):
             ordered=False,
         )
 
+    def test_union_with_different_models(self):
+        expected_result = {
+            "Angel Di Maria",
+            "Lionel Messi",
+            "Emiliano Martinez",
+            "Gabriel Marques",
+            "Jorge Luis Borges",
+            "Umberto Eco",
+        }
+        Celebrity.objects.create(name="Angel Di Maria")
+        Celebrity.objects.create(name="Lionel Messi")
+        Celebrity.objects.create(name="Emiliano Martinez")
+        Celebrity.objects.create(name="Gabriel Marques")
+        Number.objects.filter(num__lte=1).values(alias=F("other_num"))
+        e1 = ExtraInfo.objects.create(value=7, info="e1")
+        Author.objects.create(name="Gabriel Marques", num=1, extra=e1)
+        Author.objects.create(name="Jorge Luis Borges", num=2, extra=e1)
+        Author.objects.create(name="Umberto Eco", num=3, extra=e1)
+        qs1 = Celebrity.objects.values(alias=F("name"))
+        qs2 = Author.objects.values(alias_author=F("name"))
+        qs_union = qs1.union(qs2).values("name").all()
+        self.assertEqual(set(e["name"] for e in qs_union), expected_result)
```

Comment:
If `set()` is needed to compare order-insensitively (and not remove duplicates), you can use `assertCountEqual()`.
```diff
diff --git a/tests/queries/test_qs_combinators.py b/tests/queries/test_qs_combinators.py
index 9bee332599..f674c7b419 100644
--- a/tests/queries/test_qs_combinators.py
+++ b/tests/queries/test_qs_combinators.py
@@ -143,12 +143,12 @@ class QuerySetSetOperationTests(TestCase):
         Author.objects.create(name="Umberto Eco", num=3, extra=e1)
         qs1 = Celebrity.objects.values(alias=F("name"))
         qs2 = Author.objects.values(alias_author=F("name"))
-        qs_union = qs1.union(qs2).values("name").all()
-        self.assertEqual(set(e["name"] for e in qs_union), expected_result)
-        qs_union = qs1.union(qs2).all()
-        self.assertEqual(set(e["alias"] for e in qs_union), expected_result)
-        qs_union = qs2.union(qs1).all()
-        self.assertEqual(set(e["alias_author"] for e in qs_union), expected_result)
+        qs3 = qs1.union(qs2).values("name")
+        self.assertCountEqual([e["name"] for e in qs3], expected_result)
+        qs4 = qs1.union(qs2)
+        self.assertCountEqual([e["alias"] for e in qs4], expected_result)
+        qs5 = qs2.union(qs1)
+        self.assertCountEqual([e["alias_author"] for e in qs5], expected_result)
 
     @skipUnlessDBFeature("supports_select_intersection")
     def test_intersection_with_empty_qs(self):
```